**GEOG5415M Programming for Spatial Data Science**

# Week 3: Using and exploring data - data cleaning and `for` loops

**Contact: N.S.Malleson@leeds.ac.uk**

## Importing Packages

Python has a lot of basic functionality built in, but a lot of the time while doing data analysis you will require more than the basic functionality. This is when you need to import packages.

Sometimes these packages come pre-installed with python and you just need to tell python that you want to use them. The `pandas` package, which we'll use a lot in this course, is one of these. But other times you have to install them yourself beforehand.

If you have seen anyone else's code before you will note people tend to install all packages at the beginning of their code. This i

To import a package use the statement <code>import</code> followed by the <code>package_name</code>.

<br> <font color='orchid'> <b>Run the code below</b></font> to import the pandas package. <br>(nothing will appear in the cell but a number will appear in the squared brackets to show that the code has finished running)



In [1]:
import pandas

### The pandas package
The pandas package allows us to easily create and handle dataframes, similar to Excel. Data are usually put into an easily readable format of columns and rows called a `DataFrame`.

It also allows us to read in data from a CSV file or excle file.

To tell python when we want to use a function that is part of a package, you need to put the package name first. For example we would use `pandas.read_csv()` to call the `read_csv()` function.

**Note on abbreviating packages**

To save us having to type out `pandas` every time, we can abbreviate `pandas` to `pd` using the code bellow. `pd` is a common convention, but we could use any string.

<font color='orchid'> <b>Run the code below</b></font>.

This is often done for commonly used packages to save time and reduce the likelihood of typos while coding.

So <code>pandas.read_csv()</code> becomes <code>pd.read_csv()</code>.

In [2]:
import pandas as pd

 Now let's load some of the other packages we will need for the rest of the exercise

<font color='orchid'> <b>Run the code below</b></font>

In [3]:
# Import other required packages

# packages for visualising data
import matplotlib.pyplot as plt
import seaborn as sns

## Load data for the tutorial

Rather than downloading and loading our own dataset, here we will use a useful dataset that is available in the `seaborn` package. It contains information about the passengers who were on the Titanic

<font color='orchid'> <b>Run the chunks below</b></font> to read the data and check it is as we were expecting

In [4]:
# load the titanic example dataset (from the seaborn package) and assign it as a dataframe named titanic
titanic = sns.load_dataset('titanic')

The titanic dataset should be a pandas `DataFrame`:

In [5]:
print(type(titanic))

<class 'pandas.core.frame.DataFrame'>


To check that the data have been loaded correctly and are in the format that we were expecting, lets have a look at the first few rows:

In [6]:
# look at the first 5 rows of the dataframe
titanic.head(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


## Summary statistics

We can also explore the different columns and get some basic summary statisitcs by using the function `.describe()`. This function is available on all `DataFrame` objects.

<br> <font color='orchid'> <b>Run the code below</b></font> (Notice that it is only possible to summarise the details of columns that have a numeric (interger or float) data type).


In [7]:
titanic.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


:question: **Question**: Why do you think most columns have 891 values, but `age` only has 714?

### Summarising a single column
<code>.describe()</code> can also be used on just one column:

<b>Remember:</b> The easiest way to refer to a column is by putting the name of the column in square brackets and speech marks [" "] after the name of the dataframe. e.g.

```python
dataframe_name["column_name"].describe()
```

We can also do that for multiple columns if you put them in a list, like:

```python
dataframe_name[ ["column_name1", "column_name2"] ].describe()
```
(note the extra square bracket because we are passing a _list_ of columns, not just a single column).

<b> <font color='orchid'> <b> Write code to describe the "age" column of the titanic dataframe </b></font>

In [8]:
# Describe the "age" column
titanic['age'].describe()

,age
count,714.000000
mean,29.699118
std,14.526497
min,0.420000
25%,20.125000
50%,28.000000
75%,38.000000
max,80.000000


Or on a subset of columns.

<br> <font color='orchid'> <b> Write code below to describe the "age" and "survived" columns of the titanic dataframe </b></font>

In [9]:
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [12]:
# Describe the "age" and |survived" columns simultaneously
titanic_sub=titanic[['survived','age']]
titanic_sub.describe()

,survived,age
count,891.000000,714.000000
mean,0.383838,29.699118
std,0.486592,14.526497
min,0.000000,0.420000
25%,0.000000,20.125000
50%,0.000000,28.000000
75%,1.000000,38.000000
max,1.000000,80.000000


In [14]:
titanic[['survived','age']].describe()


,survived,age
count,891.000000,714.000000
mean,0.383838,29.699118
std,0.486592,14.526497
min,0.000000,0.420000
25%,0.000000,20.125000
50%,0.000000,28.000000
75%,1.000000,38.000000
max,1.000000,80.000000


### Mean, median, mode, quantile, std, count
We can also just get one of the metrics from the <code>describe()</code> function e.g. <code>.mean()</code>, <code>.quantile(0.75)</code>, <code>.max()</code>

Use the age column of the titanic data to explore these functions in the cells below.

#### Mean
<code>.mean()</code>
<br> <font color='orchid'> <b> Run the code below to calculate the mean "age" </b></font>

In [ ]:
titanic['age'].mean()

#### Median
<code>.median()</code>
<br> <font color='orchid'> <b> Run the code below to calculate the median "age" </b></font>

In [ ]:
titanic['age'].median()

#### Quantile
<code>.quantile(q)</code><br>
Where: <br>
q: Quantile or sequence of quantiles to compute, which must be between 0 and 1 inclusive.

<br> <font color='orchid'> <b> Run the code below to calculate quantiles for the "age" variable </b></font>

In [ ]:
# Get the 75% quantile
titanic['age'].quantile(0.75)

<br> <font color='orchid'> <b>Write your own code below to get the 25% quantile</b></font>

In [ ]:
# Get the 25% quantile


#### Maximum and Minimum

<code>.max()</code> <code>.min()</code>
<br> <font color='orchid'> <b> Run the code below to get the oldest passenger</b></font>

In [ ]:
# get the oldest passenger
titanic['age'].max()

<br> <font color='orchid'> <b>Write your own code below to get the youngest passenger</b></font>

In [ ]:
# get the youngest passenger


#### n smallest and largest

<code>nsmallest(n=)</code><code>nlargest(n=)</code>

Sometimes it is useful to be able to look at n number of values at the extreme of the data. This can especially be useful to identify outliers in the data.

<br> <font color='orchid'> <b>Write your own code below to identify the 10 youngest and 5 oldest passenger</b></font>

In [ ]:
# 10 youngest passengers


In [ ]:
# 5 oldest passengers


## Reading in some coffee Data

That's all for the Titanic now. We're going to start with some different data for the rest of the practical

Those of you who made it to the extra task will have already read in some 'coffee' data from a CSV file using Pandas. Here we go into more detail about what you were doing.

#### About the data <a class="tocSkip">   

I have fabricated some data about buying coffee (my favourite thing to do) and coffee prices close to the University. We will use this data in parts of the exercise later on. Before we get started with any analysis we need to donload the data and load it in to pandas.

##### Data format <a class="tocSkip">   

The data is originally in a CSV (comma separated variable) format. Most data in excel can be saved as a CSV and commonly downloaded data comes as a CSV file. A nice thing about CSV files is that they are plain text, so you can look at them using a basic text editor (like Notepad on Windows or TextEdit on Mac). This is sometimes useful for checking the data before loading it in to pandas.
    
An example of what a CSV file looks like. The first line usually has the column headers and the remaining lines have the data

<img src="https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/blob/main/notebooks/screenshots/csv_screenshot.png?raw=1" width="600" />

### Reading in CSV files

To read the data into a dataframe we will use the pandas package introduced above, using the <code>pd.read_csv()</code> function.


The <code>pd.read_csv()</code> function requires a 'filepath', i.e. we need to tell python where to find the data. To do this we put the filepath in speechmarks within the brackets of the function.
I.e. <code>pd.read_csv("<font color='blue'>file_path</font>")</code>
Often this path points to a file on our computer, but it can also be a URL that points to a file on the Internet.

#### File paths
File paths are realtive to where this notebook is saved. Our file structure looks like this:

```
GEOG5415M-Programming-for-Spatial-Data-Science/
├── data/
│   ├── week_1/
│   ├── week_2/
│   ├── week_3/
│   |   ├── coffee_data.csv
│   |   ├── london_merged.csv
│   └── week_x/
├── notebooks/
│   ├── Week_1_Introduction_to_python.ipynb
│   ├── Week_2_Python_data_types.ipynb
│   ├── Week_3_Using_exploring_and_cleaning_data.ipynb
│   └── Week_X...
└── README.md
```

As this notebook is in the `notebooks` folder, to get to the c`coffee_data.csv` file we need to:
 1. Go 'up' one directory (that would put us in `GEOG5415M-Programming-for-Spatial-Data-Science/`). To move 'up' a directory we use two dots `..`
 2. Move into `data`
 3. Then into `week3`

Therefore the path is: `../data/week_3/coffee_data.csv` (ask a demonstrator if you're not sure about this).

There is a <a href='#file_paths'>quick note about file paths</a> below if these are new to you.

#### URLS

To read in data from a URL on github we need to copy and paste the link to the 'raw data'. As the coffee data are on github, we can use:

`https://raw.githubusercontent.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/refs/heads/main/data/week_3/coffee_data.csv`.

Notice '/raw/refs/' in the URL. This is so we get the raw data, not the github page [about the data](https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/blob/main/data/week_3/coffee_data.csv).

Too get the raw link righ-click copy link on the 'raw' button on the csv file in Github:

![getting a link to raw data](https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/blob/main/screenshots/Week3_raw_csv.png?raw=1)


<br> <font color='orchid'> <b>Run the code below</b></font> to read the data from the URL

In [ ]:
url = 'https://raw.githubusercontent.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/refs/heads/main/data/week_3/coffee_data.csv'

# Open csv file in a Pandas Dataframe
coffee= pd.read_csv(url)
# view first 5 rows of the coffee datafame
coffee.head()

In [ ]:
# OR, go to https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/blob/main/data/week_3/coffee_data.csv
# and click  the 'download raw file' button in the left hand corner
# you can then upload the file using the link generated by the code below

# Uncomment to upload the file to colab:
#from google.colab import files
#files.upload()

In [ ]:
# if you have used the 'goole.colab import files' option uncomment this code to upload coffee_data
#coffee=pd.read_csv('coffee_data.csv')

##### So what have we done in the above code?

We have told pandas to read in a csv file and provided the path to where the CSV file can be found.

We have also told pandas to assign the dataframe the name `coffee`.

It is useful to have logical dataframe names  rather than df1, df2, etc. So you know what you are referring to when you come back to your code.

## Task
Using the code from this week and last week, find out the following information about the coffee data (answers at the end of the section)
- Using <code>.describe()</code>, which variables have summary statistics shown? Why only these variables?
- What is the average (mean) price of coffee?
- What is the standard deviation in coffee price?
- What is the median coffee rating?
- Using <code>.nlargest()</code> how many coffees scored 5/5?

<br> <font color='orchid'> <b>Write your own code to answer the questions in the chunks below</b></font>

In [ ]:
# Use .describe() to find summary statistics for the coffee variables. Which variables are shown, and why only these variables?


In [ ]:
# What is the average (mean) price of coffee?


In [ ]:
# What is the standard deviation in coffee price?


In [ ]:
# What is the median coffee rating?


In [ ]:
# Using .nlargest() how many coffees scored 5/5?


Number of coffees scoring 5/5: (type answer here: ... )

## Data Cleaning

Data cleaning is the process of replacing, modifying, or deleting records from a set of data that have been identified as incomplete, incorrect, inaccurate or irrelevant.

### Checking for missing data

Often the first step in data cleaning is to check for missing data. Earlier on we used <code>.info()</code> to get an overview of the titanic data frame and identify the number of non-null values in each columns. Remember <code>NaN</code> denotes a cell that contains no data (a null object).

<br> <font color='orchid'> <b>Re-run <code>titanic.info()</code> to remind yourself</b></font>

In [15]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


Equally we can used <code>.isna()</code> or <code>.isnull()</code> which both have the same function; to identify if the cell in the dataframe contains no data.

<br> <font color='orchid'> <b>Run both sets of code below to see which rows contain null data.</b></font>

In [16]:
titanic.isna()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
887,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,True,False,False,False,False,False,False,False,True,False,False,False
889,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [17]:
titanic.isnull()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False
887,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,True,False,False,False,False,False,False,False,True,False,False,False
889,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False


This shows us row by row if the value is null or not, however it would also be useful to know the total number of null rows. To do this simply add <code>.sum()</code> after <code>.isna()</code> or <code>.isnull()</code>.

In [18]:
titanic.isnull().sum()

,0
survived,0
pclass,0
sex,0
age,177
sibsp,0
parch,0
fare,0
embarked,2
class,0
who,0


From this we can see a few passengers are missing age data, 2 are missing data on where they embarked but a lot are missing data on which deck they were on

There are many ways to deal with missing data and it is up to you how you choose to handle missing values. For the sake of this exercise I will show you a few methods.

#### Drop unwanted columns

As so many passengers are missing data on which deck they are on we are going to remove or 'drop' the deck column from the titanic dataframe.

To do this we use the <code>.drop()</code> function.
We have to specify:
- the lable of the column (or index) we want to drop <code>['deck']</code>
- the axis to be dropped: whether to drop labels from the index (0 or ‘index’) or columns (1 or ‘columns’).

Full information about the <code>.drop</code> function can be found here https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html

Here are the three ways to drop the 'deck' column using the `drop` function.

<font color='orchid'> <b>Run the chunks to see what they do.</b></font> Have a look at the resulting tables, has the column definitely gone?

In [ ]:
# 1
titanic.drop(['deck'], axis=1)

In [20]:
# 2
titanic.drop(['deck'],axis='columns')

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


In [19]:
# 3
titanic.drop(columns =['deck'])

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


However, if we now check the titanic columns, 'deck' is still there!

In [21]:
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

This is because, as with most pandas functions, when we call the `drop` function it doesn't actually make changes to our original titanic dataframe. Instead, it makes a new dataframe that doesn't include the deck column. In the tables you saw above (the ones without the 'deck' column), those were the _new_ dataframes created by `drop`. So, to keep those changes, we need to replace the orginial titnaic dataframe with the new titanic dataframe created by `drop`.

To do this we simply need to write
```python
titanic = titanic.drop(columns =['deck'])
```
(i.e. create a new dataframe that happens to be called `titanic` and will therefore override the old one).

In [22]:
titanic = titanic.drop(columns =['deck'])
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

Now the titanic dataframe does not contain the 'deck' column.

Be careful to check the code has done what you want it to before when redefining dataframes as you may accidentally loose data or information. If you want to check something first, before overriding your dataframe, you could assign the new dataframe a different name, for example:
```python
temporary_titanic_df = titanic.drop(columns =['deck'])
```
and then later, once you're sure it's OK, override the original with:
```python
titanic = temporary_titanic_df
```

### Dropping unwanted rows

In the case of the passengers with missing embarking inforamtion we may want to remove these two passengers form the dataframe.

#### Droppping null rows

Similar to <code>.drop()</code> and <code>.isna()</code> we can used <code>.dropna()</code> which drops rows with null values in a particualr column e.g.

<code>titanic['embarked'].dropna()</code> drops rows where 'embarked' is NaN (although this isn't very helpful on its own as well see).

<font color='orchid'> <b>Run code below to see what `dropna` does.</b></font>  What type of data does this return?

In [23]:
titanic['embarked'].dropna()

,embarked
0,S
1,C
2,S
3,S
4,S
...,...
886,S
887,S
888,S
889,C


### Using .loc to access groups of rows and or columns

Rather than just dropping the rows where 'embarked' is NaN, sometimes it is more helpful to look at the rows first. We can use `loc` to do this.

The <code><font color='blue'>dataframe_name</font>.loc[]</code> attribute of pandas dataframes allows us to access rows and/or columns of the dataframe we want to edit or view.

Recall that the basic syntax for `loc` is:
```python
  df.loc[ <rows_to_filter>, <columns_to_select> ]
```
and if you only want to select rows or columns, not both, you can use a colon as follows:
```python
  df.loc[ <rows_to_filter> , : ]     # Filter some rows and return all columns
  df.loc[ : , <columns_to_select> ]  # Return all rows, but only some columns
```

You can also be lazy with `loc`, so if you just want to filter out some rows and return all columns, you can leave off the comma and colon:
```python
  df.loc[ <rows_to_select> ]
```

<br><font color='orchid'><b> Run through the next few cells and make sure you understand how <code>.loc</code> is working to select the different columns and rows </b></font>

The following will only show rows where 'embarked' is NaN:

In [24]:
titanic.loc[titanic['embarked'].isna(), :]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
61,1,1,female,38.0,0,0,80.0,NaN,First,woman,False,NaN,yes,True
829,1,1,female,62.0,0,0,80.0,NaN,First,woman,False,NaN,yes,True


which is exactly the same as this (note the comma and colon have gone)

In [ ]:
titanic.loc[titanic['embarked'].isna()]

We can also look at rows where the column contains a certain attribute or attributes. For example we can see which passengers embarked in Southampton using <code>titanic.loc[titanic['embark_town']=='Southampton']</code>

<b>Note:</b>
- the double equals sign == is used to compare the values (whereas a single = is used to assign a value to a variable)
- Southampton is a string so we use ' '

In [ ]:
titanic.loc[titanic['embark_town']=='Southampton']

<code>.loc[]</code> can also be used for numbers. E.g. find where the fare paid is £30

In [ ]:
titanic.loc[titanic['fare']==30]

For numeric columns we can also use other comaprison operators other than <code> == </code>

![Python comparison operators](https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/blob/main/notebooks/screenshots/Python-Comparison-Operators.jpg?raw=1) Source: https://data-flair.training/blogs/python-comparison-operators/

<font color='orchid'> <b>Write code using the `loc` function to answer the following questions:</b></font>

- How many passengers paid less than £30? (note that number of rows = number of passengers)
- How many passengers were male?
- How many passengers were 75 or older?
- How many passengers did not embark in Southampton?

In [ ]:
# How many passengers paid less than £30 (number of rows = number of passengers)?


In [ ]:
# How many passengers were male (number of rows = number of passengers)?


In [ ]:
# How many passengers were 75 or older?


In [ ]:
# How many passengers did not embark in Southampton?


Finally, we might just be interested in a variable dependent on another variable value. E.g. The boarding class of passengers who paid over £100.

In [ ]:
titanic.loc[titanic['fare']>=100, 'class']

### Rounding data

The youngest passenger in the titanic data is 0.42 years old (~aprox. 5 months). However, the age of all passengers over 1 is rounded to the nearest year.

We can clean the data so that all participant ages are to the nearest year. Wecan do this using the <code>.round(n)</code> function, where n is the number of decimal places

In [ ]:
# first identify passengers under 1
titanic.loc[titanic['age']<1]

In [ ]:
# round age to whole number (0 decimal places)
titanic['age'].round(0)

That gives us a list of ages, rounded to 0 decimal places. But we don't _do_ anything with that list other than display it. If we want to keep those rounded numbers then we need to assign them back to the original titanic dataset (similar to what we did when we used `drop`):

In [ ]:
titanic['age'] = titanic['age'].round(0)

In [ ]:
# re-identify passengers under 1
titanic.loc[titanic['age']<1]

## For loops with more than one variable

Recall that a `for` loop allows us to iterate over a list. We create a temporary variable that represents each list item. For example in the following loop, we use the variable `i`:

In [ ]:
for i in [1, 2, 3, 4, 5]:
    print(i)

In the list above, each element is a single number. But the elements in the lists can be any type of object, including **collections of objects**. For example, we could create a list that contains other lists:

```python
l = [ ['a',1], ['b',2], ['c',3] ]
```
Question, what will happen if we run the code below? <font color='orchid'> Think about this and then <b>Run the code below.</b></font>

In [ ]:
l = [ ['a',1], ['b',2], ['c',3] ]
for i in l:
    print(i)

As you can see, each time we iterate, our `i` variable is a list.

In many cases we might want to acceess those list items and do something with them. We can use _list indexing_ (which we learned about last week) to do that like this:

In [ ]:
l = [ ['a',1], ['b',2], ['c',3] ]
for i in l:
    print("The letter is:", i[0], "and the number is:", i[1])

But there's another handy way to use the for loop. Rather than having a single variable, `i`, that gets assigned to a sub-list, we can have python _unpack_ the sub-lists for us automatically but declaring _two_ variables in the loop. The code above is equivalent to.

In [ ]:
l = [ ['a',1], ['b',2], ['c',3] ]
for x, y in l:
    print("The letter is:", x, "and the number is:", y)

See how we're looping over two variables, `x` abnd `y` at the same time? We will use that feature a lot in pandas

## `iterrows()` method

The `iterrows()` method is a great example of why it can be useful to iterate over two values at once. Have a quick look at the [`iterrows() documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iterrows.html)]. It **yields** two things: the _index_ of a dataframe, and a _series_ (which is like a pandas dataframe, but only 1 row) of all the items in the corresponding row. So, we can use our handy for loops to run over these two variables at the same time.

<br> <font color='orchid'> <b>Run the code below to work out what is going on</b></font>

In [ ]:
for index, row in titanic.iterrows():
    print("Row number:", index, "Person alive?", row['alive'], "Person travel class:", row['class'])

<br> <font color='orchid'> <b>Edit this markdown cell to describe what's going on above</b></font>

...

### If-Else statements:
If-else statements allow us ot run a block of code only if certain conditions are met.
<br> <font color='orchid'> <b>Run the following two blocks of code</b></font> and see if you can work out how if-else statements work.

In [ ]:
# These examples just use 'if'

my_name = "Nick"

if my_name == "Nick":
  print("Hello, Nick")

if my_name == "Ava":
  print("Hello, Ava")

In [ ]:
# These examples use if and else
# (else is used if the 'if' is not true)

my_name = "Nick"

if my_name == "Nick":
  print("Hello, Nick")
else:
  print("I don't know your name")

if my_name == "Ava":
  print("Hello, Ava")
else:
  print("I still don't know your name")

The below code uses the iterrows() method from above. Then uses an if statement. If the row contains a passenger who is alive, male and in third class it prints the row number and the string 'lucky'. As we can see the number fo items printed it a lot shorter than the total number of passengers as 'lucky' is only printed when the row meets those conditions.

<br> <font color='orchid'> <b>Run the code below </b></font>

In [ ]:
for index, row in titanic.iterrows():
    if (row['alive']=='yes') & (row['sex']=='male') & (row['pclass']==3):
        print(index, 'lucky')

Adding an 'else' statement allows us to add a blcok of code if the 'if' conditionis False

<br> <font color='orchid'> <b>Run the code below </b></font>

In [ ]:
for index, row in titanic.iterrows():
    if (row['alive']=='yes') & (row['sex']=='male') & (row['pclass']==3):
        print(index, 'lucky')
    else:
        print(index, 'unlucky')

You will notice that by adding an else statement the number of printed items increases. As passengers can either meet the if condition *(lucky)* or not *(unlucky)*.  So the number of pritned statements will be the same as the number fo passengers in the titanic df.

Try adding more conditions using elif (else-if)
<br> <font color='orchid'><b>Add an elif statement to the statement above based on another set of titanic charactersitcs.</b></font> \
Use the following link as guideance https://www.programiz.com/python-programming/if-elif-else

In [ ]:
for index, row in titanic.iterrows():
    if (row['alive']=='yes') & (row['sex']=='male') & (row['pclass']==3):
        print('lucky')
    elif # add your code here  :

    else:
        print('unlucky')

### Back to the coffee data

We will also look at some data cleaning using the coffee data frame (introduced in the extra task).

In [ ]:
coffee.describe()

It seems that £20.00 probably has the decimal in the wrong place. £2.00 seems a much more sensible price for coffee. This is an example of a fairly common error, particualrly if the data has been entered by a human.

There are several ways we can do this

1) Re-wrtie the list using <code>[]</code> with each variable seperated by a comma.
    We are going to assign the cleaned coffee variables to a new column called 'price_clean1' so we can compare columns later.
    
(This is pretty time consuming and likely to introduce more errors)
<br> <font color='orchid'> <b>Run the code below</b></font>

In [ ]:
# rewrite the coffee as a list with the correct value in place
# and use = to asign this to coffee['price_clean1']
coffee['price_clean1']= [2.4, 2.6, 2.5, 2.0, 2.5, 2.7, 2.75, 3.0, 2.9, 1.8, 3.0, 2.5, 3.0, 2.7, 2.2, 2.0, 1.0, 2.85, 1.9, 3.1, 2.1, 2.4, 2.4, 0.8]
# print to check
print(coffee['price_clean1'])

In [ ]:
# create another 2 duplicates of column price and name them price_clean2, price_clean3
# we will use them later on
coffee['price_clean2'] = coffee['price']
coffee['price_clean3'] = coffee['price']
# printing the columns shows they are duplicates of the original 'price' column
print(coffee['price_clean2'],coffee['price_clean3'])

Typing out the whole list of coffee prices is quite time consuming, we wouldn't be able to do it if we had 100s of rows of data.
Instead we can identify which row the incorrect value is in and correct just that value.

In [ ]:
print(coffee['price'])

We can see that the £20.00 coffee is in row 15. Note that the row indexing starts at zero (not one) in Python.

We can use the <code><font color='blue'>dataframe_name</font>.iloc</code> attribute of pandas dataframes to access rows and/or columns of the dataframe we want to edit.  

The first item in the squared brackets identifies the row(s) of interest and the second item (after the comma) identifies the column(s) of interest
<code><font color='blue'>dataframe_name</font>.iloc[<font color='blue'>row_number(s)</font>,<font color='blue'>column_number(s)</font>]</code>

NOTE: to select all rows or all columns use a colon <code>:</code>

<br><font color='orchid'><b> Run through the next few cells and make sure you understand how <code>.loc</code> and <code>iloc</code> are working to select the different columns and rows </b></font>

In [ ]:
# select row 5 and all columns
coffee.iloc[[5],:]

In [ ]:
# select row 3-7 and columns 'price' and 'rating'
coffee.loc[[3,4,5,6,7],['price','rating']]

In [ ]:
# using iloc we can use the index of the columns as well.
# select rows 0-3 and columns 0-4
coffee.iloc[[0,1,2,3],[0,1,2,3,4]]

In [ ]:
# check the location of the incorrect data is row 15 in column price_clean2
coffee.loc[[15],"price_clean2"]

Repeat the above using .iloc[] (note 'price_clean2' is column 7)

In [ ]:
# wrtie code here


In [ ]:
# assign the incorrect data the correct value
coffee.loc[[15],"price_clean2"] = 2.0
print(coffee['price_clean2'])

The above two methods are great if only a few values need cleaning or the dataset is small enough to easily find the location of the incorrect data.
However for large datasets these methods are impracticle and applying a more general rule would be better.

We can hypothesis that any coffee over about £8 probably has had the price incorrectly entered and that the decimal point should be moved. Therefore we can find every row where the price is over £8 and divide it by 10 (effectively moving the decimal place back a digit).


<br> <font color='orchid'> <b>Run the code below</b></font>

Locate where coffee price is greater than or equal to £8: <code>coffee.loc[coffee['price']>=8] </code>

In [ ]:
# is coffee price greater than or equal to £8.
# gives a boolean (true/false) answer
coffee['price']>=8

In [ ]:
# use .loc to identify the row where coffee price is >= £8
coffee.loc[coffee['price']>=8]

Finally we combine those operations to clean all the rows that have unreasonably high prices, and store the new (hopefully correct) price in the column 'price_clean3'.

<font color='orchid'> <b>Run the code bellow, using the comments to work out what each bit of the code is doing. Importnat: ask if you have any questions! </b></font> The code _looks_ complicated, but it is easy to explain.

In [ ]:
# in rows where coffee price is >= £8,
# in column 'price_clean3' calculate the cleaned price of coffee
# by dividing the original coffee['price'] by 10

coffee.loc[coffee['price']>=8, 'price_clean3'] = coffee['price']/10
# view column price_clean3
print(coffee['price_clean3'])

### Let's check our cleaning
Lets check and compare our cleaning methods
We can see the original price of £20.00 and then how each method has cleaned the data to £2.00
<br> <font color='orchid'> <b>Run the code below</b></font>

In [ ]:
# show row 15 of the datrframe and the price columns
coffee.loc[[15],['price','price_clean1','price_clean2','price_clean3']]

## Answers
![answers_coffee_2.png](https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/blob/main/notebooks/screenshots/answers_coffee_2.png?raw=1)

## Bike share data
For this exercise we will be using bike share data from Transport for London, downloaded from [Kaggle](https://www.kaggle.com/hmavrodiev/london-bike-sharing-dataset). I have downloaded the data and saved it as a csv file called 'london_merged.csv' in the data directory. You can also download the dataset yourselves.

#### Note about Kaggle <a class="tocSkip">
Kaggle is a great source of datasets which are free to download to get started with doing some data analysis. They also run data challenges where a dataset and problem will be given for you to solve. These challenges are a great way to practice your data science skills. However, I do not recommend using Kaggle to source data for assessments. As lots of people have analysed the data before - there is greater temptation to plagiarise code (this is really easy for the lecturers to spot).

#### About the data <a class="tocSkip">
    
Bike sharing is where bicycles are made available for shared use to individuals on a short term basis for a price or free. Many bike share systems allow people to borrow a bike from a "dock" and return it at another dock belonging to the same system. Docks are special bike racks that lock the bike, and only release it by computer control. The user enters payment information, and the computer unlocks a bike. The user returns the bike by placing it in the dock, which locks it in place. [Source](https://en.wikipedia.org/wiki/Bicycle-sharing_system)
    
If you are unfamiliar with the concept of bike sharing in the UK have a quick read of this [article](https://tfl.gov.uk/modes/cycling/santander-cycles/how-it-works) <br>

Some metadata information about the variables:
- <b>cnt</b>: The count of new bike shares
- <b>timestamp</b>: date and time of new bike share
- <b>t1</b>: real temperature in Celsius
- <b>t2</b>: "feels like" temperature in Celsius
- <b>hum</b>: humidity in percentage
- <b>wind_speed</b>: wind speed in km/h
- <b>weather_code</b>: category of the weather
    - Description:
        - 1 = Clear ; mostly clear but have some values with haze/fog/patches of fog/ fog in vicinity
        - 2 = scattered clouds / few clouds
        - 3 = Broken clouds
        - 4 = Cloudy
        - 7 = Rain/ light Rain shower/ Light rain
        - 10 = rain with thunderstorm
- <b>is_holiday</b>: boolean field: 1 holiday / 0 non holiday
- <b>is_weekend</b>: boolean field: 1 if the day is weekend/ 0 weekday
- <b>season</b>: category field of meteorological seasons: 0: spring ; 1: summer; 2: autumn; 3:winter
    
<br> <font color='orchid'> <b>Run the code below to read in the data </b></font>



In [ ]:
# have a look at how we read in the coffee data and see if yoou can write your own code to read the london_merged.csv in
# hint, the url for the raw csv is:
# https://github.com/MSc-Urban-Environmental-Leeds/GEOG5415M-Programming-for-Spatial-Data-Science/raw/refs/heads/main/data/week_3/london_merged.csv

bike = pd.read_csv()

#### Explore the data <a class="tocSkip">
Explore the data yourselves using the <code>.head()</code>, <code>.tail()</code> and <code>.describe()</code> functions introduced in exercise 1.
<br> <font color='orchid'> <b>Write your own code below </b></font>

In [ ]:
# explore the data


In [ ]:
# explore the data


Let us also check the data type of each of the columns in the bike dataframe using <code>.dtypes</code> to see if they are what we expect.
<br> <font color='orchid'> <b>Run the code below </b></font>

In [ ]:
bike.dtypes

#### Metadata and data type comaprison  <a class='tocSkip'>

We would expect temperature (t1,t2), percentage humidity (hum) and wind speed (wind_speed) to be floats as these values are not limited to whole whole numbers. <br>

We would also expect cnt to be an integer as you cannot get half a bike share. <br>

Though is_holiday and is_weekend are boolean (1 or 0 aka True or False) we will leave them as integers for now to make later analysis easier.<br>

We can see that the timestamp column has been read in as object. While it should be a datetime data type. We can convert the object to datetime using the pandas function <code>pd.to_datatime</code>

<br> <font color='orchid'> <b>Run the code below </b></font>

In [ ]:
bike['timestamp']

#### Time series data

Have a look at the code below, what does "%d/%m/%Y %H:%M" stand for? (more info about datetime formats here https://www.programiz.com/python-programming/datetime/strftime

In [ ]:
bike['timestamp'] = pd.to_datetime(bike['timestamp'], format="%d/%m/%Y %H:%M")

### Have a go at adding extra columns that show the month, day of the month, hour, and day of the week.

Hint: <code> bike['timestamp'].year </code> gives the year of each timestamp, use the pandas datetime documentation and google e.g. https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.day.html

In [ ]:
bike['year'] = bike['timestamp'].dt.year
bike['year'].head()

In [ ]:
# finish this line of code
bike['month']=

In [ ]:
# finish this line of code
bike['day_of_month']=

In [ ]:
# finish this line of code
bike['hour']=

In [ ]:
# finish this line of code
bike['day_of_week']=

In [ ]:
# re-run d.types to check timestamp is now datetime
bike.dtypes

## Introduction to the  groupby function
### What is the<code>.groupby()</code> function?
An explanation from the [package documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html):<br>
<i>"A groupby operation involves some combination of splitting the object, applying a function, and combining the results. This can be used to group large amounts of data and compute operations on these groups"</i>

<code>.groupby()</code> is useful to group data to identify interesting patterns and calculate some summary statistics that give us more insight into the data. <code>.groupby()</code> can also be used to group data before we plot it, to make more interesting graphs and visualisations.

We can group the bike dataframe by <code>'is_weekend'</code>, then use the <code>.size()</code> function to calculate how many days are classified as being a weekend or weekday.
<br> <font color='orchid'> <b>Run the code below </b></font>

In [ ]:
# group the bike dataframe by 'is_weekend', then use the .size() function
bike.groupby('is_weekend').size()

We can also group the data and then calculate summary statistics (using the functions introduced last week: <code>.mean()</code>, <code>.median()</code>, <code>.quantile(0.25)</code>, <code>.max()</code>, <code>.min()</code>)

<br> <font color='orchid'> <b>Run the code below then experiment calculating your own summary statistics on grouped data</b></font>

In [ ]:
bike.groupby('is_weekend').mean()

In [ ]:
# experiment here


In [ ]:
# experiment here


In [ ]:
# experiment here


If you are interested in calculating the summary stats. of one column you can put it in squared brackets after the groupby function <br> I.e. <code><font color='red'>dataframe_name</font>.groupby("<font color='red'>column_name to groupby</font>")["<font color='red'>column_name to calc. summary stats for</font>"].mean()</code>

In [ ]:
# group the bike dataframe by 'is_weekend' and calcualte the mean number of bike shares ('cnt') for weekdays and weekends
bike.groupby('is_weekend')['cnt'].mean()

<font color='orchid'><b> Write your own function to group the bike dataframe by 'is_weekend' and calculate the <b>maximum</b> daily number of bike shares ('cnt') on a weekday and weekend </font> (Solution at the end of the exercise if you get stuck)

In [ ]:
bike.groupby('is_weekend')['cnt'].max()

<font color='orchid'> <b>Write your own function to calculate the mean number of bike shares for each season </b> </font> (Solution at the end of the exercise if you get stuck)

In [ ]:
bike.groupby('season')['cnt'].mean()

<font color='orchid'> <b>Write your own function to calculate the minimum number of daily bike shares for each weather type </b></font> (Solution at the end of the exercise if you get stuck)

In [ ]:
bike.groupby('weather_code')['cnt'].min()

Let us calculate the total number of bike share uses per day. Currently the bike share data is provided for every hour of every day so we need to group these together to calculate the totals for the day. First of all we need to create a new column for just the date. <br>
We do this using the datetime.date function: <code>.dt.date</code>
<br> <font color='orchid'> <b>Run the code below </b></font>

In [ ]:
bike['date']= bike['timestamp'].dt.date

In [ ]:
# let's check it has worked
bike.head()

Next we need to use the <code>.groupby()</code> function to get a dataframe where each row is a unqiue day.<br>

To do this we need to think carefully about how we want to aggregate the columns of the dataframe:<br>

For <b>count</b> it makes sense to take the total number of bike share uses across all hours to get the total for the day: therefore we want to aggregate using the <code>.sum()</code> function.<br>

For the <b>temperature, humidity and wind speed</b> variables it does not make sense to take the sum of all hours (we would end up with very high temeperatures, humidities and wind speeds) instead taking the average value of all hours of the day makes more sense, using the <code>.mean()</code> function. <br>

As <b>weather_code</b> is a non-continuous categorical variable, where a number is used to represent the weather conditions, taking the most common hourly weather type during a day is the most sensible way to accurately represent the overall weather for that day. We can do this using the <code>pd.Series.mode</code> function.

For the <b> holiday, weekend, season</b> and <b>date</b> variables they will be the same across all hours of the day, so no calculation is needed. Instead we will just take the first value (for ease) using the <code>.first()</code> function.
<br> <font color='orchid'> <b>Run the code below to test this out </b></font>

In [ ]:
# let's test this out
bike.groupby('date')['cnt'].sum()

In [ ]:
# let's test this out
# we can add .head() to show just the first 5 lines
bike.groupby('date')[['t1','t2','hum','wind_speed']].mean().head()

In [ ]:
# let's test this out
bike.groupby('date')['weather_code'].apply(pd.Series.mode).head()

In [ ]:
# let's test this out
bike.groupby('date')[['is_holiday','is_weekend','season','date']].first().head()

We can do this all separately, but what if we want to produce a dataframe of the aggregated by day bike data?
It is not just as easy as directly combining the above code.
<font color='orchid'> Uncomment (remove #) and run the code below, note the error message </font>

In [ ]:
#bike.groupby('date')['t1','t2','hum','wind_speed'].mean()['is_holiday','is_weekend','season','date'].first()

To solve this we can use <code>.agg({})</code> function to apply ["one or more operations over the specified axis"](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.agg.html).
<br> For each column we specify which function we want to apply to the aggreagated rows.<br>
The <code>{}</code> indicate a dictionary object where a set of objects (keys) are mapped to another set of objects (values)
i.e. the column is mapped to the required aggregation method.

<font color='orchid'> Run the code below and check you understand what it is doing </font>

In [ ]:
bike_day = bike.groupby('date').agg({'cnt':'sum', 't1':'mean','t2':'mean','hum':'mean',
                                     'wind_speed':'mean','weather_code':pd.Series.mode,
                                     'is_holiday':'first','is_weekend':'first',
                                     'season':'first','date':'first'})
bike_day.head()

We will revisit this tomorrow when we look at data visualisation and statistical modelling

# Extra Content

Some useful further practice:
- for loops: https://holypython.com/intermediate-python-exercises/exercise-8-python-for-loop/
- if-else statements: https://holypython.com/intermediate-python-exercises/exercise-7-conditional-statements/
- using groupby https://www.w3resource.com/python-exercises/pandas/groupby/index.php#EDITOR
- dealing with missing values https://www.w3resource.com/python-exercises/pandas/missing-values/index.php)
- querying data https://www.w3resource.com/python-exercises/pandas/movies/index.php

## FAQs

## Why do some functions have parentheses and others do not? <a class="tocSkip">

Parentheses indicate the difference between methods versus attibutes

### Attributes: <a class="tocSkip">
- do not have parenthesis ()
- are values assocaited with an object e.g. <code> <font color='blue'>dataframe</font>.shape, <font color='blue'>dataframe</font>.column, <font color='blue'>dataframe</font>.dtypes </code> these are all attrributes of the dataframe (object) they are appliead to.

### Methods: <a class="tocSkip">
- do have partnethesis ()
- are functions assocaited with particular objects. E.g. <code> <font color='blue'>dataframe</font>.head(), <font color='blue'>dataframe</font>["<font color='blue'>column</font>"].mean() </code>